src/sparse_matrix_builder.rs

```diff
@@ -0,0 +1,276 @@
+        }
+    }
+}
+
+impl SparseMatrixDescriptor {
+    pub fn new(col_a_id: u8, col_a_name: String, col_b_id: u8, col_b_name: String) -> Self {
+        Self {
+            col_a_id,
+            col_a_name,
+            col_b_id,
+            col_b_name,
+        }
+    }
+
+    pub fn make_buffer(&self) -> SparseMatrixBuffer {
+        SparseMatrixBuffer {
+            descriptor: self.clone(),
+            edge_count: 0,
+            hash_2_row: Default::default(),
+            hashes_2_edge: Default::default(),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct SparseMatrixBuffer {
+    pub descriptor: SparseMatrixDescriptor,
+    pub edge_count: u32,
+    hash_2_row: HashMap<u64, Row, BuildHasherDefault<FxHasher>>,
+    hashes_2_edge: HashMap<(u64, u64), f32, BuildHasherDefault<FxHasher>>,
+}
+
+impl SparseMatrixBuffer {
+    pub fn handle_pair(&mut self, hashes: &[u64]) {
+        let a = self.descriptor.col_a_id;
+        let b = self.descriptor.col_b_id;
+        self.add_pair_symmetric(
+            hashes[(a + 1) as usize],
+            hashes[(b + 1) as usize],
+            hashes[0],
+        );
+    }
+
+    fn add_pair_symmetric(&mut self, a_hash: u64, b_hash: u64, count: u64) {
+        let value = 1f32 / (count as f32);
+
+        self.update_row(a_hash, value);
+        self.update_row(b_hash, value);
+
+        self.edge_count += 1;
+
+        self.update_edge(a_hash, b_hash, value);
+        self.update_edge(b_hash, a_hash, value);
+    }
+
+    fn update_row(&mut self, hash: u64, val: f32) {
+        let mut e = self.hash_2_row.entry(hash).or_default();
+        e.occurrence += 1;
+        e.row_sum += val;
+    }
+
+    fn update_edge(&mut self, a_hash: u64, b_hash: u64, val: f32) {
+        let e = self.hashes_2_edge.entry((a_hash, b_hash)).or_default();
+        *e += val;
+    }
+}
+
+#[derive(Debug)]
+pub struct SparseMatrixBuffersReducer {
+    descriptor: SparseMatrixDescriptor,
+    buffers: Vec<SparseMatrixBuffer>,
+    node_indexer: NodeIndexer,
+}
+
+impl SparseMatrixBuffersReducer {
+    pub fn new(node_indexer: NodeIndexer, buffers: Vec<SparseMatrixBuffer>) -> Self {
+        if buffers.is_empty() {
+            panic!("Cannot reduce 0 buffers")
+        }
+
+        let descriptor = buffers[0].descriptor.clone();
+        for buffer in &buffers {
+            if descriptor != buffer.descriptor {
+                panic!("Can only reduce buffers with the same sparse matrix description")
+            }
+        }
+
+        Self {
+            descriptor,
+            buffers,
+            node_indexer,
+        }
+    }
+
+    pub fn reduce(self) -> SparseMatrix {
+        let rows = self.reduce_row_maps();
+        let hashes_2_edge = self.reduce_edge_maps();
+
+        let mut entities: Vec<_> = {
+            hashes_2_edge
+                .into_par_iter()
+                .map(|((row_hash, col_hash), edge)| {
+                    let row = *self.node_indexer.key_2_index.get(&row_hash).unwrap() as u32;
```

Comment:
Add comment stating that these entries must be in the hashmap

Reply:
added 'expect'